aiida_quantumespresso/calculations/pwimmigrant.py

```diff
@@ -115,6 +117,13 @@ class or using the ``set_`` methods (e.g. ``set_remote_workdir``).
             absolute path to the job's directory on that computer.
         :type remote_workdir: str
 
+        :param calc_hooks: one or more of the input nodes can be linked with a function which
+        takes as input those input nodes and returns them (in the same order).
+        In this way the input nodes can be linked to precursor Calculation node(s).
+        Allowed keys are "structure", "kpoints", "parameters", "settings" and can only be used once
+        an example would be calc_hooks=[[my_func1, ["structure", "parameters"]], [my_func2, ["kpoints"]]]
```

Comment:
I am not sure I understand the intended behavior of these hooks. When I take the example usage detailed in the PR message, you want to call a function that will be passed the `StructureData` that is generated from the completed calculation that is to be imported, and that function should then serve to connect that `StructureData` to a `CifData`, the latter which should be retrieved from the structure input. However, I don't see how that is possible, since the `StructureData` that will be passed is the reconstructed one from the completed calculation that we are importing and so will not yet exist in the graph. How should that be linked to an existing `CifData`? Similarly, in the documentation here you describe that these hooks should serve to link up the to be created/imported input nodes to previously existing calculations. However, the hook will be generic, correct? If you do not pass in the actual calculation, how is the hook going to determine which calculation to link to the input nodes, based on those input nodes? Should the calculation that the user wants to link everything to, not be passed as an input to the hook itself?

Reply:
So yes its a little complex lol. What I do is create the function as a closure.
This is what it looks like in my code:

```python
for sname in ["pyrite"]:
    for modname in ["AFM", "FM", "NM"]:     

       ... # setup calc with specific inputs

        cif = get_cif_node(sname, known_cod_ids.get(sname, None))
        # ParameterData which informs how to modify the CIF structure 
        # e.g. making a primitive cell and setting spin configuration
        cif_mod = get_cifmod_node(sname, modname)

        def cifhook(struct, params):

            class Cif2structDummy(WorkChain):
                @classmethod
                def define(cls, spec):
                    super(Cif2structDummy, cls).define(spec)
                    spec.input('cif')
                    spec.input('cif_mod')
                    spec.outline(
                        cls.results,
                    )
                    spec.output('struct')
                    spec.output('params')

                def results(self):
                    self.out('struct', struct)
                    params.label = modname
                    self.out('params', params)

            Cif2structDummy.run(cif=cif, cif_mod=cif_mod)

            calc.create_input_nodes(open_transport, calc_hooks=[[cifhook, ["structure", "parameters"]]])
```

and this is an example of the resulting graph:

![example_image_tree](https://user-images.githubusercontent.com/2997570/44521904-af562e00-a6cc-11e8-9ed8-ea35448a3c19.png)

Reply:
This is what has worked for me, but open to suggestions.

On a side note, do you know if it is possible to set a default label and/or description in a `WorkChain` subclass?